### PR TITLE
Draft how-to guide for storage

### PR DESCRIPTION
**Pull request by tonyandrewmeyer**
To add to Discourse for juju.is/docs/sdk in the "how-to"/"develop" section.

### CODE REVIEW

**Comment by tonyandrewmeyer** [comment]
@tmihoc would you be able to give this a review?

File: sdk/how-to-guides/manage-storage.md [hunk 1]
```diff
@@ -0,0 +1,232 @@
+> See first: [Juju | How to Manage Storage](https://juju.is/docs/juju/manage-storage)
```
Comment (tmihoc) [inline sdk/how-to-guides/manage-storage.md]: Please use sentence case.

File: sdk/how-to-guides/manage-storage.md [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+
+**Contents:**
+
+- [Implement storage management](#heading--implement-storage-management)
```
Comment (tmihoc) [inline sdk/how-to-guides/manage-storage.md]: Why "storage management"? Actually I think I understand why, but it's clunky. Thinking...

File: sdk/how-to-guides/manage-storage.md [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+
+<a href="#heading--implement-storage-management"><h2 id="heading--implement-storage-management">Implement storage management</h2></a>
+
+- [Declare the storage in `charmcraft.yaml`](#heading--declare-the-storage-in-charmcraft-yaml)
+- [Observe the attached event and define an event handler](#heading--observe-the-attached-event-and-define-an-event-handler)
+
+
+<a href="#heading--declare-the-storage-in-charmcraft-yaml"><h3 id="heading--declare-the-storage-in-charmcraft-yaml">Declare the storage in `charmcraft.yaml`</h3></a>
+
+To define the storage that can be provided to the charm, define a `storage` section in `charmcraft.yaml` that lists the storages and information about each storage. For example, for a transient filesystem storage mounted to `/cache/` that is at least 1GB in size:
```
Comment (tmihoc) [inline sdk/how-to-guides/manage-storage.md]: Storage volumes?

File: sdk/how-to-guides/manage-storage.md [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+  local-cache:
+      type: filesystem
+      description: Somewhere to cache files locally.
+      location: /cache/
+      minimum-size: 1G
+      properties:
+          - transient
+```
+
+> See more: [File `charmcraft.yaml` > `actions`](/t/7132#heading--storage)
```
Comment (tmihoc) [inline sdk/how-to-guides/manage-storage.md]: typo: > `storage`

File: sdk/how-to-guides/manage-storage.md [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+
+<a href="#heading--observe-the-attached-event-and-define-an-event-handler"><h3 id="heading--observe-the-attached-event-and-define-an-event-handler">Observe the attached event and define an event handler</h3></a>
+
+In the `src/charm.py` file, in the `__init__` function of your charm, set up an observer for the attached event associated with your storage and pair that with an event handler, typically a holistic one. For example:
```
Comment (tmihoc) [inline sdk/how-to-guides/manage-storage.md]: I'd remove "attached" (it's not clear to me what it is adding).

File: sdk/how-to-guides/manage-storage.md [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+
+```
+self.framework.observe(self.on.cache_storage_attached, self._update_configuration)
+```
+
+> See more: [StorageAttachedEvent](https://ops.readthedocs.io/en/latest/#ops.StorageAttachedEvent), [Juju SDK | Holistic vs delta charms](https://juju.is/docs/sdk/holistic-vs-delta-charms)
```
Comment (tmihoc) [inline sdk/how-to-guides/manage-storage.md]: Here and everywhere else: Please use code formatting and the `ops.` prefix.

File: sdk/how-to-guides/manage-storage.md [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+
+Now, in the body of the charm definition, define the event handler, or adjust an existing holistic one. For example, providing the location of the attached storage to the workload configuration:
```
Comment (tmihoc) [inline sdk/how-to-guides/manage-storage.md]: typo: to provide*

File: sdk/how-to-guides/manage-storage.md [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+        # Pebble isn't ready yet. Return and wait for the pebble-ready event.
+        logger.info("Pebble is not yet ready.")
+        return
+```
+
+> Examples: [ZooKeeper ensuring that permission and ownership is correct](https://github.com/canonical/zookeeper-operator/blob/106f9c2cd9408a172b0e93f741d8c9f860c4c38e/src/charm.py#L247), [Kafka configuring additional storage](https://github.com/canonical/kafka-k8s-operator/blob/25cc5dd87bc2246c38fc511ac9c52f35f75f6513/src/charm.py#L298)
+
+<a href="#heading--observe-the-detaching-event-and-define-an-event-handler"><h3 id="heading--observe-the-detaching-event-and-define-an-event-handler">Observe the detaching event and define an event handler</h3></a>
+
+In the `src/charm.py` file, in the `__init__` function of your charm, set up an observer for the detacting event associated with your storage and pair that with an event handler. For example:
```
Comment (tmihoc) [inline sdk/how-to-guides/manage-storage.md]: detecting*? (What's a detecting event?)

File: sdk/how-to-guides/manage-storage.md [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+
+```
+def _on_storage_detaching(self, event: ops.StorageDetachingEvent):
+    """Handle the storage being detached."""
+    self.unit.status = ops.ActiveStatus("Caching disabled; provide storage to boost performance)
+```
+
+> Examples: [MySQL handling cluster management](https://github.com/canonical/mysql-k8s-operator/blob/4c575b478b7ae2a28b09dde9cade2d3370dd4db6/src/charm.py#L823), [MongoDB updating the set before storage is removed](https://github.com/canonical/mongodb-operator/blob/b33d036173f47c68823e08a9f03189dc534d38dc/src/charm.py#L596)
+
+<a href="#heading--requesting-additional-storage"><h2 id="heading--requesting-additional-storage">Requesting additional storage</h2></a>
```
Comment (tmihoc) [inline sdk/how-to-guides/manage-storage.md]: Request*

File: sdk/how-to-guides/manage-storage.md [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+```
+
+For example, if the charm needs to request two additional units of this storage:
+
+```python
+self.model.storages.request("scratch", 2)
+```
+
+The storage will not be available immediately after that call - the charm should
+observe the `storage-attached` event (as described above) and handle any remaining
```
Comment (tmihoc) [inline sdk/how-to-guides/manage-storage.md]: I'd avoid references of the form "as described above". They lead to entanglement.

File: sdk/how-to-guides/manage-storage.md [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+
+- [Write unit tests](#heading--write-unit-tests)
+- [Write scenario tests](#heading--write-scenario-tests)
+- [Write integration tests](#heading--write-integration-tests)
+
+<a href="#heading--write-unit-tests"><h3 id="heading--write-unit-tests">Write unit tests</h3></a>
+
+> See first: [How to write unit tests for a charm](/t/4461)
+
+When using Harness for unit tests, use the `add_storage()` method to simulate Juju adding storage to the charm. You can either have the method also simulate attaching the storage, or do that explicitly with the `attach_storage()` method. In this example, we verify that the charm responds as expected to storage attached and detaching events:
```
Comment (tmihoc) [inline sdk/how-to-guides/manage-storage.md:147]: Using Harness as opposed to what?

File: sdk/how-to-guides/manage-storage.md [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+> See first: [How to write scenario tests for a charm](/t/10585)
+
+When using Scenario for unit tests, to verify that the charm state is as expected after storage changes, use the `run` method of the Scenario `Context` object. For example, to provide the charm with mock storage:
+
+```python
+# Some charm with a 'foo' filesystem-type storage defined in its metadata:
+ctx = scenario.Context(MyCharm)
+storage = scenario.Storage("foo")
+
+# Setup storage with some content:
```
Comment (tmihoc) [inline sdk/how-to-guides/manage-storage.md]: Set up*

File: sdk/how-to-guides/manage-storage.md [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+    myfile.write_text("helloworlds")
+
+# Verify that the contents are as expected afterwards.
+assert (
+    storage.get_filesystem(ctx) / "path.py"
+).read_text() == "helloworlds"
+```
+
+If a charm requests adding more storage instances while handling some event, you
+can inspect that from the Context.requested_storage API.
```
Comment (tmihoc) [inline sdk/how-to-guides/manage-storage.md]: Use code formatting.

File: sdk/how-to-guides/manage-storage.md [hunk 14]
```diff
@@ -0,0 +1,232 @@
...
+
+```python
+ctx = scenario.Context(MyCharm)
+ctx.run('some-event-that-will-request-more-storage', scenario.State())
+
+# The charm has requested two 'foo' storages to be provisioned:
+assert ctx.requested_storages['foo'] == 2
+```
+
+Requesting storages has no other consequence in Scenario. In real life, this
```
Comment (tmihoc) [inline sdk/how-to-guides/manage-storage.md]: Is "storages" okay? I've never heard it in the plural. I'm guessing it's understood as storage volumes, but even then -- is it common enough to use this way?

File: sdk/how-to-guides/manage-storage.md (hunk 5 above)
Comment (tonyandrewmeyer) [reply sdk/how-to-guides/manage-storage.md]: To distinguish between the attached and detaching events. Maybe 'storage-attached'?

File: sdk/how-to-guides/manage-storage.md (hunk 8 above)
Comment (tonyandrewmeyer) [reply sdk/how-to-guides/manage-storage.md]: detaching :)

File: sdk/how-to-guides/manage-storage.md (hunk 11 above)
Comment (tonyandrewmeyer) [reply sdk/how-to-guides/manage-storage.md:147]: Using Scenario.

File: sdk/how-to-guides/manage-storage.md (hunk 3 above)
Comment (tonyandrewmeyer) [reply sdk/how-to-guides/manage-storage.md]: Looking over the existing docs, we seem to have a mix of "storage volumes", "storage instances", and "storages", I believe all meaning the same thing.

I agree "storages" is probably not the best choice. Do you have a preference between "storage volumes" and "storage instances"? "volume" has fairly specific meaning in filesystems and in K8s, so probably "storage volumes" is clearest?

File: sdk/how-to-guides/manage-storage.md (hunk 14 above)
Comment (tonyandrewmeyer) [reply sdk/how-to-guides/manage-storage.md]: Covered under the "storage" question above - I'll adjust all the cases to match.

File: sdk/how-to-guides/manage-storage.md (hunk 3 above)
Comment (tmihoc) [reply sdk/how-to-guides/manage-storage.md]: > so probably "storage volumes" is clearest?

Yes.

**Pull request merged**